crates/ra_hir_ty/src/infer/expr.rs

```diff
@@ -63,6 +63,75 @@ impl<'a> InferenceContext<'a> {
         self.resolve_ty_as_possible(ty)
     }
 
+    fn callable_sig_from_fn_trait(&mut self, ty: &Ty) -> Option<FnSig> {
+        if let Some(krate) = self.resolver.krate() {
+            let fn_traits: Vec<crate::TraitId> = [FnTrait::FnOnce, FnTrait::FnMut, FnTrait::Fn]
+                .iter()
+                .filter_map(|f| get_fn_trait(self.db, krate, *f))
+                .collect();
+            for fn_trait in fn_traits {
+                let fn_trait_data = self.db.trait_data(fn_trait);
+                let generic_params = generics(self.db.upcast(), fn_trait.into());
+                if generic_params.len() != 2 {
+                    continue;
+                }
+
+                let arg_ty = self.table.new_type_var();
+                let substs = Substs::build_for_generics(&generic_params)
+                    .push(ty.clone())
+                    .push(arg_ty.clone())
+                    .build();
+
+                let trait_ref = TraitRef { trait_: fn_trait, substs: substs.clone() };
+                let trait_env = Arc::clone(&self.trait_env);
+                let implements_fn_goal =
+                    self.canonicalizer().canonicalize_obligation(InEnvironment {
+                        value: Obligation::Trait(trait_ref),
+                        environment: trait_env,
+                    });
+                if let Some(Solution::Unique(SolutionVariables(solution))) =
+                    self.db.trait_solve(krate, implements_fn_goal.value.clone())
+                {
+                    match solution.value.as_slice() {
+                        [Ty::Apply(ApplicationTy {
+                            ctor: TypeCtor::Tuple { cardinality: _ },
+                            parameters,
+                        })] => {
+                            let output_assoc_type = match fn_trait_data
+                                .associated_types()
+                                .collect::<Vec<hir_def::TypeAliasId>>()
+                                .as_slice()
+                            {
+                                [output] => *output,
+                                _ => {
+                                    continue;
+                                }
+                            };
+                            let output_proj_ty = crate::ProjectionTy {
+                                associated_ty: output_assoc_type,
+                                parameters: substs,
+                            };
+                            let return_ty = self.normalize_projection_ty(output_proj_ty);
+                            return Some(FnSig::from_params_and_return(
+                                parameters.into_iter().map(|ty| ty.clone()).collect(),
```

Comment:
these parameters are still canonicalized. (I.e. there might be parameters that can be anything, which are represented by bound variables in the canonicalized solution and would need to be turned into inference variables.) Fortunately, that will resolve itself with the refactorings I mention above, I think.

Reply:
I'm not following how this is resolved by the other refactorings. Should we be passing in the argument expressions for the call and somehow registering the new obligations without ever unpacking the types from the solution?

Reply:
What I mean is basically:
 - callable_sig gets passed the number of arguments (let's say 3)
 - callable_sig creates that many type variables (let's say `?1`, `?2`, `?3`)
 - callable_sig checks that the callee may implement `Fn<(?1, ?2, ?3)>`, and if so, registers that as an obligation and returns these variables `?1`, `?2`, `?3` as argument types.

Then what happens automatically is that that obligation gets checked later, and if it can be solved uniquely (or there's some definite guidance), those variables `?1` etc. get unified with the solution we get from Chalk. The way that works is that `Canonicalized::apply_solution` gets called, which correctly handles going back from the canonical solution to the inference context. 

So if you wanted to use the solution you get back here in `callable_sig`, the way to do it would be like here:
https://github.com/rust-analyzer/rust-analyzer/blob/0f7961d5570f17d6c2098ab11d2a3bcbbfb84ff6/crates/ra_hir_ty/src/infer.rs#L331-L337
but as mentioned, that would just be an optimization since it happens automatically later anyway.

Reply:
To give an example of why this works this way, let's say we have some obligation `SomeType: Fn<(?1,)>`, and an impl like `impl<T> Fn<(Option<T>,)> for SomeType`. So Chalk determines that `?1 = Option<... anything ...>`, but there's no restriction on what the 'anything' can be. The way that looks in the solution is that we get back a `Canonical` with one binder, and the solution is `Option<^0.0>`; so the `^0.0` is a bound variable referring to that binder of the canonical. This means that we need to introduce a new type variable `?2`; and then we have `?1 = Option<?2>` and `?2` can be anything. So "applying the solution" involves potentially learning the values of the original type variables, but also introducing new ones.

Reply:
Ok, I think I've got it now.

Reply:
I made these changes, but a couple other things came up:
1. We still need to run `trait_solve` and make sure that the answer is `Some` so that we pick the right type from autoderef chain. I'm wondering if something could go wrong if we accept ambiguous solutions: is it possible that one type in the autoderef chain doesn't implement `FnOnce`, but `trait_solve` returns `Ambig::Unknown` so we push the wrong obligation? 

1. Actually I guess it's ok to just check the `FnOnce` trait instead of going through all three. I think they need to check all three in rustc because they need more than just the argument and return types.

Reply:
1. Yeah, but it works the same way in rustc, so it should be fine.
2. Yes. If we do things like move checking we'll need to be more precise, but for now checking for `FnOnce` is actually enough.